File: bindings/r/R/safe_version.R
# SPDX-License-Identifier: PMPL-1.0
# SPDX-FileCopyrightText: 2025 Hyperpolymath

#' Safe semantic versioning parsing and comparison.
#'
#' Provides SemVer parsing with validation and safe comparison
#' operations following the Semantic Versioning 2.0.0 specification.

#' Create a Version object.
#'
#' @param major Major version number
#' @param minor Minor version number
#' @param patch Patch version number
#' @param prerelease Optional prerelease string
#' @param build_metadata Optional build metadata string
#' @return Version S3 object
#' @keywords internal
new_version <- function(major, minor, patch, prerelease = NULL, build_metadata = NULL) {
  structure(
    list(
      major = as.integer(major),
      minor = as.integer(minor),
      patch = as.integer(patch),
      prerelease = prerelease,
      build_metadata = build_metadata
    ),
    class = "Version"
  )
}

#' Check if object is a Version.
#'
#' @param x Object to check
#' @return TRUE if x is a Version object
#' @export
is_version <- function(x) {
  inherits(x, "Version")
}

#' Format Version object.
#'
#' @param x Version object
#' @param ... Additional arguments (ignored)
#' @return Formatted version string
#' @export
format.Version <- function(x, ...) {
  version_to_string(x)
}

#' Print Version object.
#'
#' @param x Version object
#' @param ... Additional arguments (ignored)
#' @export
print.Version <- function(x, ...) {
  cat("Version:", version_to_string(x), "\n")
  invisible(x)
}

#' Create a new version.
#'
#' @param major Major version number
#' @param minor Minor version number
#' @param patch Patch version number
#' @return Version object
#' @export
version <- function(major, minor, patch) {
  new_version(major, minor, patch)
}

#' Add prerelease to version.
#'
#' @param ver Version object
#' @param prerelease Prerelease string
#' @return New Version object with prerelease
#' @export
version_with_prerelease <- function(ver, prerelease) {
  if (!is_version(ver)) return(NULL)
  new_version(ver$major, ver$minor, ver$patch, prerelease, ver$build_metadata)
}

#' Add build metadata to version.
#'
#' @param ver Version object
#' @param build Build metadata string
#' @return New Version object with build metadata
#' @export
version_with_build <- function(ver, build) {
  if (!is_version(ver)) return(NULL)
  new_version(ver$major, ver$minor, ver$patch, ver$prerelease, build)
}

#' Parse a version string.
#'
#' @param version_string Version string (e.g., "1.2.3-alpha+build")
#' @return Version object or NULL on parse error
#' @export
parse_version <- function(version_string) {
  if (is.na(version_string) || is.null(version_string)) {
    return(NULL)
  }

  s <- trimws(version_string)

  # Remove optional 'v' prefix
  if (startsWith(tolower(s), "v")) {
    s <- substr(s, 2, nchar(s))
  }

  # Split off build metadata
  build_metadata <- NULL
  plus_pos <- regexpr("+", s, fixed = TRUE)
  if (plus_pos > 0) {
    build_metadata <- substr(s, plus_pos + 1, nchar(s))
    s <- substr(s, 1, plus_pos - 1)
  }

  # Split off prerelease
  prerelease <- NULL
  dash_pos <- regexpr("-", s, fixed = TRUE)
  if (dash_pos > 0) {
    prerelease <- substr(s, dash_pos + 1, nchar(s))
    s <- substr(s, 1, dash_pos - 1)
  }

  # Parse major.minor.patch
  parts <- strsplit(s, ".", fixed = TRUE)[[1]]
  if (length(parts) != 3) {
    return(NULL)
  }

  major <- suppressWarnings(as.integer(parts[1]))
  minor <- suppressWarnings(as.integer(parts[2]))
  patch <- suppressWarnings(as.integer(parts[3]))

  if (is.na(major) || is.na(minor) || is.na(patch)) {
    return(NULL)
  }

  if (major < 0 || minor < 0 || patch < 0) {
    return(NULL)
  }

  new_version(major, minor, patch, prerelease, build_metadata)
}

#' Convert version to string.
#'
#' @param ver Version object
#' @return Version string or NA
#' @export
version_to_string <- function(ver) {
  if (is.null(ver) || !is_version(ver)) {
    return(NA_character_)
  }

  s <- sprintf("%d.%d.%d", ver$major, ver$minor, ver$patch)

  if (!is.null(ver$prerelease)) {
    s <- paste0(s, "-", ver$prerelease)
  }

  if (!is.null(ver$build_metadata)) {
    s <- paste0(s, "+", ver$build_metadata)
  }

  s
}

#' Check if version is a prerelease.
#'
#' @param ver Version object or string
#' @return TRUE if prerelease
#' @export
version_is_prerelease <- function(ver) {
  if (is.character(ver)) {
    ver <- parse_version(ver)
  }

  if (is.null(ver) || !is_version(ver)) {
    return(NA)
  }

  !is.null(ver$prerelease)
}

#' Check if version is stable (>= 1.0.0 and no prerelease).
#'
#' @param ver Version object or string
#' @return TRUE if stable
#' @export
version_is_stable <- function(ver) {
  if (is.character(ver)) {
    ver <- parse_version(ver)
  }

  if (is.null(ver) || !is_version(ver)) {
    return(NA)
  }

  ver$major >= 1 && is.null(ver$prerelease)
}

#' Increment major version.
#'
#' @param ver Version object
#' @return New Version object
#' @export
version_bump_major <- function(ver) {
  if (!is_version(ver)) return(NULL)
  new_version(ver$major + 1, 0, 0)
}

#' Increment minor version.
#'
#' @param ver Version object
#' @return New Version object
#' @export
version_bump_minor <- function(ver) {
  if (!is_version(ver)) return(NULL)
  new_version(ver$major, ver$minor + 1, 0)
}

#' Increment patch version.
#'
#' @param ver Version object
#' @return New Version object
#' @export
version_bump_patch <- function(ver) {
  if (!is_version(ver)) return(NULL)
  new_version(ver$major, ver$minor, ver$patch + 1)
}

#' Compare two versions.
#'
#' @param ver1 First Version object or string
#' @param ver2 Second Version object or string
#' @return -1 if ver1 < ver2, 0 if equal, 1 if ver1 > ver2, NA if invalid
#' @export
version_compare <- function(ver1, ver2) {
  if (is.character(ver1)) ver1 <- parse_version(ver1)
  if (is.character(ver2)) ver2 <- parse_version(ver2)

  if (is.null(ver1) || is.null(ver2) || !is_version(ver1) || !is_version(ver2)) {
    return(NA_integer_)
  }

  # Compare major.minor.patch
  if (ver1$major != ver2$major) {
    return(if (ver1$major < ver2$major) -1L else 1L)
  }

  if (ver1$minor != ver2$minor) {
    return(if (ver1$minor < ver2$minor) -1L else 1L)
  }

  if (ver1$patch != ver2$patch) {
    return(if (ver1$patch < ver2$patch) -1L else 1L)
  }

  # Prerelease comparison
  pre1 <- ver1$prerelease
  pre2 <- ver2$prerelease

  if (is.null(pre1) && is.null(pre2)) return(0L)
  if (is.null(pre1)) return(1L)  # No prerelease > prerelease
  if (is.null(pre2)) return(-1L)

  # Compare prerelease strings
  if (pre1 == pre2) return(0L)
  if (pre1 < pre2) return(-1L)
  return(1L)
}

#' Check if version satisfies a constraint.
#'
#' Supports: >=, <=, >, <, =, ^, ~ operators.
#'
#' @param ver Version object or string
#' @param constraint Constraint string (e.g., ">=1.0.0", "^1.2.0")
#' @return TRUE if satisfies, FALSE otherwise, NA on error
#' @export
version_satisfies <- function(ver, constraint) {
  if (is.character(ver)) ver <- parse_version(ver)
  if (is.null(ver) || !is_version(ver)) return(NA)

  constraint <- trimws(constraint)

  # Parse operator and target version
  if (startsWith(constraint, ">=")) {
    target <- parse_version(substr(constraint, 3, nchar(constraint)))
    if (is.null(target)) return(NA)
    return(version_compare(ver, target) >= 0)
  }

  if (startsWith(constraint, "<=")) {
    target <- parse_version(substr(constraint, 3, nchar(constraint)))
    if (is.null(target)) return(NA)
    return(version_compare(ver, target) <= 0)
  }

  if (startsWith(constraint, ">") && !startsWith(constraint, ">=")) {
    target <- parse_version(substr(constraint, 2, nchar(constraint)))
    if (is.null(target)) return(NA)
    return(version_compare(ver, target) > 0)
  }

  if (startsWith(constraint, "<") && !startsWith(constraint, "<=")) {
    target <- parse_version(substr(constraint, 2, nchar(constraint)))
    if (is.null(target)) return(NA)
    return(version_compare(ver, target) < 0)
  }

  if (startsWith(constraint, "=")) {
    target <- parse_version(substr(constraint, 2, nchar(constraint)))
    if (is.null(target)) return(NA)
    return(version_compare(ver, target) == 0)
  }

  if (startsWith(constraint, "^")) {
    # Caret: compatible with version (same major, if major > 0)
    target <- parse_version(substr(constraint, 2, nchar(constraint)))
    if (is.null(target)) return(NA)

    if (target$major == 0) {
      return(ver$major == 0 && ver$minor == target$minor &&
             version_compare(ver, target) >= 0)
    }
    return(ver$major == target$major && version_compare(ver, target) >= 0)
  }

  if (startsWith(constraint, "~")) {
    # Tilde: same major.minor
    target <- parse_version(substr(constraint, 2, nchar(constraint)))
    if (is.null(target)) return(NA)
    return(ver$major == target$major && ver$minor == target$minor &&
           version_compare(ver, target) >= 0)
  }

  # Exact match
  target <- parse_version(constraint)
  if (is.null(target)) return(NA)
  version_compare(ver, target) == 0
}

#' Equality operator for Version objects.
#'
#' @param e1 First Version
#' @param e2 Second Version
#' @return TRUE if equal
#' @export
`==.Version` <- function(e1, e2) {
  version_compare(e1, e2) == 0
}

#' Less than operator for Version objects.
#'
#' @param e1 First Version
#' @param e2 Second Version
#' @return TRUE if e1 < e2
#' @export
`<.Version` <- function(e1, e2) {
  version_compare(e1, e2) < 0
}

#' Greater than operator for Version objects.
#'
#' @param e1 First Version
#' @param e2 Second Version
#' @return TRUE if e1 > e2
#' @export
`>.Version` <- function(e1, e2) {
  version_compare(e1, e2) > 0
}

File: bindings/typescript/src/safe_version.ts
// SPDX-License-Identifier: AGPL-3.0-or-later
// SPDX-FileCopyrightText: 2025 Hyperpolymath

export interface Result<T> {
  ok: boolean;
  value?: T;
  error?: string;
}

/**
 * SemVer represents a semantic version.
 */
export class SemVer {
  readonly major: number;
  readonly minor: number;
  readonly patch: number;
  readonly prerelease: string[];
  readonly build: string[];

  private constructor(
    major: number,
    minor: number,
    patch: number,
    prerelease: string[] = [],
    build: string[] = []
  ) {
    this.major = major;
    this.minor = minor;
    this.patch = patch;
    this.prerelease = prerelease;
    this.build = build;
  }

  /**
   * Create a version.
   */
  static create(major: number, minor: number, patch: number): Result<SemVer> {
    if (major < 0 || minor < 0 || patch < 0) {
      return { ok: false, error: 'Version numbers cannot be negative' };
    }
    if (!Number.isInteger(major) || !Number.isInteger(minor) || !Number.isInteger(patch)) {
      return { ok: false, error: 'Version numbers must be integers' };
    }
    return { ok: true, value: new SemVer(major, minor, patch) };
  }

  /**
   * Parse a version string.
   */
  static parse(version: string): Result<SemVer> {
    const trimmed = version.trim();
    const withoutV = trimmed.startsWith('v') ? trimmed.slice(1) : trimmed;

    // Split build metadata
    const [versionAndPre, ...buildParts] = withoutV.split('+');
    const build = buildParts.length > 0 ? buildParts.join('+').split('.') : [];

    // Split prerelease
    const [coreVersion, ...preParts] = versionAndPre.split('-');
    const prerelease = preParts.length > 0 ? preParts.join('-').split('.') : [];

    // Parse core version
    const parts = coreVersion.split('.');
    if (parts.length < 1 || parts.length > 3) {
      return { ok: false, error: 'Invalid version format' };
    }

    const major = parseInt(parts[0], 10);
    const minor = parts.length > 1 ? parseInt(parts[1], 10) : 0;
    const patch = parts.length > 2 ? parseInt(parts[2], 10) : 0;

    if (isNaN(major) || isNaN(minor) || isNaN(patch)) {
      return { ok: false, error: 'Version parts must be numbers' };
    }

    if (major < 0 || minor < 0 || patch < 0) {
      return { ok: false, error: 'Version numbers cannot be negative' };
    }

    return { ok: true, value: new SemVer(major, minor, patch, prerelease, build) };
  }

  /**
   * Convert to string.
   */
  toString(): string {
    let result = `${this.major}.${this.minor}.${this.patch}`;
    if (this.prerelease.length > 0) {
      result += `-${this.prerelease.join('.')}`;
    }
    if (this.build.length > 0) {
      result += `+${this.build.join('.')}`;
    }
    return result;
  }

  /**
   * Compare with another version.
   * Returns -1 if this < other, 0 if equal, 1 if this > other.
   */
  compare(other: SemVer): number {
    // Compare major.minor.patch
    if (this.major !== other.major) {
      return this.major < other.major ? -1 : 1;
    }
    if (this.minor !== other.minor) {
      return this.minor < other.minor ? -1 : 1;
    }
    if (this.patch !== other.patch) {
      return this.patch < other.patch ? -1 : 1;
    }

    // Prerelease versions have lower precedence
    if (this.prerelease.length === 0 && other.prerelease.length > 0) {
      return 1;
    }
    if (this.prerelease.length > 0 && other.prerelease.length === 0) {
      return -1;
    }

    // Compare prerelease identifiers
    const minLen = Math.min(this.prerelease.length, other.prerelease.length);
    for (let i = 0; i < minLen; i++) {
      const a = this.prerelease[i];
      const b = other.prerelease[i];
      const aNum = parseInt(a, 10);
      const bNum = parseInt(b, 10);

      if (!isNaN(aNum) && !isNaN(bNum)) {
        if (aNum !== bNum) return aNum < bNum ? -1 : 1;
      } else if (!isNaN(aNum)) {
        return -1; // Numeric < alphanumeric
      } else if (!isNaN(bNum)) {
        return 1;
      } else {
        if (a < b) return -1;
        if (a > b) return 1;
      }
    }

    if (this.prerelease.length !== other.prerelease.length) {
      return this.prerelease.length < other.prerelease.length ? -1 : 1;
    }

    return 0;
  }

  /**
   * Check if equal to another version.
   */
  equals(other: SemVer): boolean {
    return this.compare(other) === 0;
  }

  /**
   * Check if greater than another version.
   */
  greaterThan(other: SemVer): boolean {
    return this.compare(other) > 0;
  }

  /**
   * Check if less than another version.
   */
  lessThan(other: SemVer): boolean {
    return this.compare(other) < 0;
  }

  /**
   * Check if greater than or equal to another version.
   */
  greaterOrEqual(other: SemVer): boolean {
    return this.compare(other) >= 0;
  }

  /**
   * Check if less than or equal to another version.
   */
  lessOrEqual(other: SemVer): boolean {
    return this.compare(other) <= 0;
  }

  /**
   * Increment major version.
   */
  bumpMajor(): SemVer {
    return new SemVer(this.major + 1, 0, 0);
  }

  /**
   * Increment minor version.
   */
  bumpMinor(): SemVer {
    return new SemVer(this.major, this.minor + 1, 0);
  }

  /**
   * Increment patch version.
   */
  bumpPatch(): SemVer {
    return new SemVer(this.major, this.minor, this.patch + 1);
  }

  /**
   * Add prerelease identifier.
   */
  withPrerelease(prerelease: string[]): SemVer {
    return new SemVer(this.major, this.minor, this.patch, prerelease, this.build);
  }

  /**
   * Add build metadata.
   */
  withBuild(build: string[]): SemVer {
    return new SemVer(this.major, this.minor, this.patch, this.prerelease, build);
  }

  /**
   * Check if this is a prerelease version.
   */
  isPrerelease(): boolean {
    return this.prerelease.length > 0;
  }

  /**
   * Check if this is a stable version (>= 1.0.0 and no prerelease).
   */
  isStable(): boolean {
    return this.major >= 1 && !this.isPrerelease();
  }

  /**
   * Check if this satisfies a version range.
   */
  satisfies(range: string): boolean {
    const ops = range.match(/^([<>=^~]*)(.+)$/);
    if (!ops) return false;

    const operator = ops[1] || '=';
    const versionStr = ops[2];
    const targetResult = SemVer.parse(versionStr);
    if (!targetResult.ok || !targetResult.value) return false;
    const target = targetResult.value;

    switch (operator) {
      case '':
      case '=':
        return this.equals(target);
      case '>':
        return this.greaterThan(target);
      case '>=':
        return this.greaterOrEqual(target);
      case '<':
        return this.lessThan(target);
      case '<=':
        return this.lessOrEqual(target);
      case '^':
        // Compatible with (same major for >= 1.0.0, same major.minor for 0.x)
        if (target.major === 0) {
          return this.major === 0 && this.minor === target.minor && this.greaterOrEqual(target);
        }
        return this.major === target.major && this.greaterOrEqual(target);
      case '~':
        // Approximately equivalent (same major.minor)
        return this.major === target.major && this.minor === target.minor && this.greaterOrEqual(target);
      default:
        return false;
    }
  }
}

/**
 * Sort versions in ascending order.
 */
export function sortVersions(versions: SemVer[]): SemVer[] {
  return [...versions].sort((a, b) => a.compare(b));
}

/**
 * Get the maximum version from a list.
 */
export function maxVersion(versions: SemVer[]): SemVer | undefined {
  if (versions.length === 0) return undefined;
  return versions.reduce((max, v) => (v.greaterThan(max) ? v : max));
}

/**
 * Get the minimum version from a list.
 */
export function minVersion(versions: SemVer[]): SemVer | undefined {
  if (versions.length === 0) return undefined;
  return versions.reduce((min, v) => (v.lessThan(min) ? v : min));
}

export const SafeVersion = {
  SemVer,
  sortVersions,
  maxVersion,
  minVersion,
};
